### PR TITLE
Reorder fw sliders

### PR DESCRIPTION
**Pull request by spatzengr**
Creating an Basic and Advanced slider dropdown so as to not overwhelm the user base with the new advanced sliders.  Also re-ordering the sliders and cleaning up some hold slider header space that was not used.
TODO:
- [x] Setup localStore for sliderLevel selection.
- [x] Finalize interaction between D_min ON vs. OFF with sliderLevelView
- [x] Look at reducing slider resolution to 0.05 increments
- [x] Get slider to persist sliderLevelView on Save
- [x] Trigger D_Min slider to 2.0 when D_min is toggled OFF.
- [x] Reivew slider reverse comptability with 4.2, 4.1, etc... and tie-in sliderLevelView to slider Enable function.
- [x] Hook sliderModeSelect to disable sliders (old behavior) when set to "OFF".

image

image

### CODE REVIEW

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:770]:
```suggestion
        if (sliderLevel === 1) {
```

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:773]: These lines are used in both cases, are they hidden otherwise?

**Review by haslinghuis — CHANGES_REQUESTED** [review]

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:773]: Oh oh - WIP - premature review ;-)

Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:773]: No, please!  I'll take all the help I can get!!!  I need some and hence why I posted it as a draft so everyone can see I'm working on something and also to make it easier to hopefully get some help to stop sitting here and staring at the code for 3 or 4 hours at a shot.  

My biggest thing right now I think is the Configurator bug(?) about it not holding variables on saves.  I need to also figure out how to pass local multiple variables to the `slideLevelView` function.  I need it to operate on the `sliderLevelSelect` dropdown and also the D_min toggle; hence why I broke out to its own function.  I reckon i just need to list both variables in each call with a comma separator?  That's where I left off last night.

Thanks in advance for your patience with me!!  Dumb-dumb brain here when it comes to this stuff, but I'm trying.  I do know some stuff, but my C++ days were 20-yrs ago and I hated that class in college.  "I know enough to be dangerous".  I wish to know more to help out.

Comment (IllusionFpv) [reply src/js/tabs/pid_tuning.js:773]: @spatzengr are you trying with firmware master branch?

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:773]: Sliders in 1.44 do work, not with older firmware.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:556]: indent here and close the }
```suggestion
                    const isDminChecked = $(this).is(':checked');
```

**Comment by mikeller** [comment]
Just having a quick look at this. I think it is perferable to have this linked to the master 'Expert mode' switch, just like all other beginner / advanced menu changes are linked to it. This is exactly the purpose of having this switch, and it was a conscious decision to not have little beginner / advanded switches on a number of different tabs.

**Comment by spatzengr** [comment]
@mikeller , here is where I'm at.  With Expert Mode, it limits the range.  The Advanced drop-down shows the new sliders.  I have a **BIG** concern that the new _advanced_ sliders - while awesome and warranted for sure - will overwhelm beginners and distract from what are the most important ratios to focus upon.  They will also give credence to those folks that love to poke at "too complicated" all the time.  With this approach, the Expert Mode toggle and the Basic/Advanced drop-down play together as you can see to limit range (Expert Mode) and also show the more advanced ratio if folks square aware the key 3 in Basic and want to take it the last 10%.  Considering that most EVERY user toggles on Expert Mode - for good reason, to check Failsafe, etc... - I don't see that warranting to show the more _Advanced_ sliders.  Honestly, 90% of any tune are the 3 Basic sliders shown in the Basic selection: PD Balance, P and D Gain and Stick Response.

I'm storing the Basic/Advanced drop-down in localStorage so if the user prefers to keep on Advanced, it stays there (on that PC of course).

image

Thoughts?

**Comment by spatzengr** [comment]
@mikeller , sorry uploaded image instead of GIF, if reviewing emails, check out GIF on PR for actions of toggles and drop-downs.

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js]: @IllusionFpv , I'm on a build from 1/5/2020 for BF version.  Once @haslinghuis fixed gets merged (#2401), I will rebase the branch on Configurator Master and flash the latest BF Master for vetting.

File: src/js/tabs/pid_tuning.js
Comment (McGiverGim) [inline src/js/tabs/pid_tuning.js:800]: I think is more simple and maintainable to add an `advancedSlider` class to all the advanced sliders, and use simple something like:

```javascript
$('.tab-pid_tuning .advancedSlider').toggle(TABS.pid_tuning.sliderLevel === 1);
if ($('.pid_tuning input[name="dMinRoll"]').val() === 0) {
    $('.tab-pid_tuning .DMinRatioSlider').hide();
}
```

It can be simplified more doing the same with all the things related to the D-Min value, but I suppose is out of this PR.

**Review by McGiverGim — CHANGES_REQUESTED** [review]

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:800]: But note the Advanced sliders are not all grouped together.  Would that still work with a class?  [thinkng emoji]

Comment (McGiverGim) [reply src/js/tabs/pid_tuning.js:800]: You need to add the same class to each slider. Is more descriptive and maintainable to have a class in each of them and add/remove the class to add/remove sliders from the advanced view that to have code with if/else/etc. that need to be modified each time.

Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:800]: OK.  Sounds good.  On it.

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:800]: Need this changes to hide `sliderRollPitchRatio`, `sliderIGain` and `sliderDMinRatio` after fixing legacy sliders.

**Comment by spatzengr** [comment]
Will revised to take out the extra dropdown and tie to Expert Mode.

image

**Comment by mikeller** [comment]
@spatzengr: Sorry for the late reply, and some of this seems to have been superseded by your poll.
I see your point about the duality coming out of using expert mode for both limited slider range and less sliders shown. However, I think 'limited slider range' should be kept for legacy configurator sliders only, and not used for the new firmware simplified settings - the problem there is that there is no equivalent for limited range when the settings are changed in OSD, or with lua on the TX, and for users it will be even more confusing if they have to deal with different ranges when using different ways to configure the same setting.
So I think keeping expert mode for limited range in the legacy configurator sliders (and keep showing all of them there), and using expert mode to show less sliders for firmware simplified settings (and adjusting OSD to have the 'always on' sliders up top, and 'expert only' sliders under a sub-heading with a descriptive title) is a viable way forward.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:800]: Added a new class firmware to hide them in legacy mode.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1863]:
```suggestion
                if (!TuningSliders.expertMode && semver.lt(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
```
Not sure what's preferred.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1007]: This block is also on line 975-981 ?

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1847]: :+1: Nice to see you work on my example

**Review by haslinghuis — CHANGES_REQUESTED** [review]

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:1863]: Agreed.  I just learned about gte vs lte vs gt vs lt.  :-)

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:1007]: good catch, thank you!

File: src/js/TuningSliders.js
Comment (haslinghuis) [inline src/js/TuningSliders.js:268]:
```suggestion
    if (this.sliderPidsMode === 2) {
```

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1965]: :+1:

**Review by haslinghuis — COMMENTED** [review]
Tripple === is preferred (look it up)

File: src/js/TuningSliders.js
Comment (McGiverGim) [inline src/js/TuningSliders.js:80]: This `if` can be totally simplified using the `toggle()` function in place of `show()` and `hide()`, as I suggested in my first review:
```javascript
const dMinShow = $('.pid_tuning input[name="dMinRoll"]').val() !== 0;
$('.tab-pid_tuning .DMinRatioSlider').toggle(this.expertMode && dMinShow);
$('.tab-pid_tuning .advancedSlider').toggle(this.expertMode);
$('.tab-pid_tuning .MasterSlider').toggle(this.expertMode);
```
And seeing that the `MasterSlider` shows and hides exactly the same than the `advancedSlider` is better simply to add the `advancedSlider` class to the `MasterSlider` and remove this toggle too.

**Review by McGiverGim — CHANGES_REQUESTED** [review]

**Comment by nangel0** [comment]
HI Everyone
Weird behavior  here using the latest nightly build available, 

fFashed latest OMNIBUSF4SD dev target with the latest configurator
the sliders went south all of them,  showing at 0, the PID values don't change when sliding.

If I perform a reset to everything the very first time I open the tab they change the values and save them without hitting the save button.

**Comment by spatzengr** [comment]
@nangel0 , make sure you are on the latest Master OR 4.2.  Below is the behavior I'm getting after flashing back and forth from Master to 4.2.8.

cKGJJKJaGw

**Comment by nangel0** [comment]
> @nangel0 , make sure you are on the latest Master OR 4.2. Below is the behavior I'm getting after flashing back and forth from Master to 4.2.8.
> 
> Note: even the same target.
> 
> cKGJJKJaGw

Thanks for the tip, but still not working, if I fash 4.2.8 from 15/02/2021 it seems ok, usual sliders change values, and the refresh button brings them back to the center if not saved, and the save button works as expected.

https://www.dropbox.com/s/guech8mk5m0pznc/4.2.8.mp4?dl=0

Flashing 4.3.0 from 18/02/2021 seems to change values but the refresh button doesn't bring to the latest values and doesn't save some changes...

https://www.dropbox.com/s/92ekl0n0uxiduqv/4.3.0.mp4?dl=0

**Comment by spatzengr** [comment]
@nangel0 , that is a build of the Master configurator and not the branch in this PR.  However, yes, in the new FW sliders, moving the sliders modifies PID direct it seems since they are not calced on the FC chip and relayed back to the Configurator so Refresh does not work the same.  Maybe that behavior can be handled differently, but outside the scope of this PR (I hope).

File: src/js/TuningSliders.js
Comment (spatzengr) [reply src/js/TuningSliders.js:80]: @McGiverGim , I get weird behavior with this method for dMinShow (see below).  :-/

MasterSlider needs to be its own class so we can call it to show for API < 1.44.

M6if0AES8s

Comment (McGiverGim) [reply src/js/TuningSliders.js:80]: Of course, your Master Slider can be shown it hide with other conditions in other places.
But if you get word behavior something is wrong with the code, toggle is exactly the same than show/hide with a if.

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:389]: This part is only available in firmware mode

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:390]: Is this causing the slider to go south?

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1918]: This part is only available in firmware mode

File: src/js/TuningSliders.js
Comment (haslinghuis) [reply src/js/TuningSliders.js:80]:
```
TuningSliders.setExpertMode = function() {
    this.expertMode = isExpertModeEnabled();
    const dMinShow = $('.pid_tuning input[name="dMinRoll"]').val() !== 0;
    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
        $('.tab-pid_tuning .DMinRatioSlider').toggle(this.expertMode && dMinShow);
        $('.tab-pid_tuning .advancedSlider').toggle(this.expertMode);
        $('.tab-pid_tuning .MasterSlider').toggle(this.expertMode);
        $('#slidersPidsBox, #slidersFilterBox').removeClass('nonExpertModeSliders');
    } else {
        if (this.expertMode) {
            $('#slidersPidsBox, #slidersFilterBox').removeClass('nonExpertModeSliders');
        } else {
            $('#slidersPidsBox, #slidersFilterBox').addClass('nonExpertModeSliders');
        }
        $('.tab-pid_tuning .advancedSlider').hide();
        $('.tab-pid_tuning .DMinRatioSlider').hide();
    }
};
```

File: src/js/TuningSliders.js
Comment (haslinghuis) [inline src/js/TuningSliders.js:270]: This part is only available for firmware sliders

**Comment by haslinghuis** [comment]
Trying to workout the problem, this is still breaking legacy sliders.

**Comment by haslinghuis** [comment]
@nangel0 In the mean time did you check the latest nightly build https://github.com/betaflight/betaflight-configurator-nightlies/releases/tag/v20210218.2008. You might have to go to defaults.

**Comment by nangel0** [comment]
> @nangel0 In the mean time did you check the latest nightly build https://github.com/betaflight/betaflight-configurator-nightlies/releases/tag/v20210218.2008. You might have to go to defaults.

Yes, I am on that version now, set up an FC just for that to stop messing with my working FCS.
The sliders seem to behave as respected, if I understand as soon as you move sliders the value are calculated and updated in the table. nice, double-clicking in the slider itself now brings the slider to center AND updates the value, this was not happening before, it only updated the value if the slide was moved to 1.

But the D_MIN button it seems to stuck no matter what.
https://www.dropbox.com/s/vavlp9dvfjtcu3g/v20210218.2008.mp4?dl=0

**Comment by haslinghuis** [comment]
@spatzengr Just tested and my suggestions will make the legacy sliders work again.

File: src/js/TuningSliders.js
Comment (haslinghuis) [reply src/js/TuningSliders.js:80]: @McGiverGim @spatzengr reverted this to your version and now it's working after making the required changes.

File: src/js/TuningSliders.js
Comment (haslinghuis) [inline src/js/TuningSliders.js:270]: Only available for firmware sliders

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by haslinghuis** [comment]
@nangel0 I don't know but thanks for testing - I found this same behavior. Only can disable Dmin switch in firmware mode when advanced is in OFF mode.

File: src/js/TuningSliders.js
Comment (haslinghuis) [inline src/js/TuningSliders.js:279]: Please only use for firmware sliders

File: src/js/TuningSliders.js
Comment (spatzengr) [reply src/js/TuningSliders.js:80]: It needs to be const` dMinShow = parseInt($('.pid_tuning input[name="dMinRoll"]').val()) !== 0;` to work.  It works now fine.

**Comment by spatzengr** [comment]
@haslinghuis @nangel0 

I made the suggested changes and pushed another commit.  Testing on 4.3 or 4.2 all appears to be in order (see below).
The D_min slider not displaying when Expert Mode is **disabled** is by design.  This is regardless if D_min is enabled (KISS).

@haslinghuis as far as my Github said, I'm in sync with the latest Master as of this moment (2/18 at 9pm EST).

**Behavior on BF 4.2.8:**
2slw1s84jj

**Behavior on current Master (BF 4.3):**
TCtqbQDSCu

File: src/js/TuningSliders.js
Comment (spatzengr) [reply src/js/TuningSliders.js]: done

File: src/js/TuningSliders.js
Comment (spatzengr) [reply src/js/TuningSliders.js]: done

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js]: done

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:390]: No and isolated now to only complete when on API 1.44 or greater.

File: src/js/tabs/pid_tuning.js
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js]: done

File: src/js/TuningSliders.js
Comment (McGiverGim) [inline src/js/TuningSliders.js:77]: For this kind of if-else, maybe is easier too use a toggle too:

```suggestion
        $('#slidersPidsBox, #slidersFilterBox').toggleClass('nonExpertModeSliders', this.expertMode);
```

All the parts when we can use a `toggle` and not an if reduces the complexity of the code.

**Review by McGiverGim — CHANGES_REQUESTED** [review]

**Comment by haslinghuis** [comment]
@spatzengr One thing remaining is keeping the sliders enabled when toggling Dmin Switch in Legacy Mode.

**Comment by spatzengr** [comment]
@haslinghuis , yeah, I'm still struggling with it.  I have copied and pasted the complete `dMinSwitch` and `TuningSliders.updatePidSlidersDisplay` code back into the files from the Master, and it still does it!!  Somehow the Master is dealing with the PID change when D_Min is toggled without treating it like a manual change in the PIDs.  However, in the PR code, it is triggering as a manual change in the PIDs, which is triggering the `TuningSliders.pidSlidersUnavailable = True`

File: src/js/tabs/pid_tuning.js
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:589]: This will fix legacy sliders
```suggestion
                if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
                    TuningSliders.updatePidSlidersDisplay();
                }
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by spatzengr** [comment]
I should clarify.  I know how to keep the legacy sliders live when toggling the D_min switch (easy).  However, as soon as `TuningSliders.updatePidSlidersDisplay` is called - if the sliders are not moved (which may be the KEY) - they shut off.  Maybe i just need to call the same function that the sliders call when they move???

**Comment by haslinghuis** [comment]
Found the solution:

```
diff --git a/src/js/TuningSliders.js b/src/js/TuningSliders.js
index d005318e..ca7b24f7 100644
--- a/src/js/TuningSliders.js
+++ b/src/js/TuningSliders.js
@@ -238,6 +238,7 @@ TuningSliders.updatePidSlidersDisplay = function() {
         pidElements.each(function (indexInput) {
             if (indexPid < 3 && indexInput < 5) {
                 if (currentPIDs[indexPid * 5 + indexInput] != $(this).val()) {
+                    console.log(`${currentPIDs[indexPid * 5 + indexInput]} ${$(this).val()}`);
                     TuningSliders.pidSlidersUnavailable = true;
                 }
                 $(this).val(currentPIDs[indexPid * 5 + indexInput]);
diff --git a/src/js/tabs/pid_tuning.js b/src/js/tabs/pid_tuning.js
index f51f6e7e..47a44b62 100644
--- a/src/js/tabs/pid_tuning.js
+++ b/src/js/tabs/pid_tuning.js
@@ -586,7 +586,9 @@ TABS.pid_tuning.initialize = function (callback) {
                     $('#pid_main .pid_titlebar2 th').attr('colspan', 5);
                     $('.derivativeText').text(i18n.getMessage("pidTuningDerivative"));
                 }
-                TuningSliders.updatePidSlidersDisplay();
+                if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
+                    TuningSliders.updatePidSlidersDisplay();
+                }
             });
             dMinSwitch.change();
         }
@@ -1815,15 +1817,16 @@ TABS.pid_tuning.initialize = function (callback) {
                 if (semver.lt(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
                     TuningSliders.setDMinFeatureEnabled($(this).is(':checked'));
                     // switch dmin and dmax values on dmin on/off if sliders available
+                    console.log('dMinSwitch');
                     if (!TuningSliders.pidSlidersUnavailable) {
                         if (TuningSliders.dMinFeatureEnabled) {
-                            ADVANCED_TUNING.dMinRoll = FC.PIDs[0][2];
-                            ADVANCED_TUNING.dMinPitch = FC.PIDs[1][2];
-                            ADVANCED_TUNING.dMinYaw = FC.PIDs[2][2];
+                            FC.ADVANCED_TUNING.dMinRoll = FC.PIDS[0][2];
+                            FC.ADVANCED_TUNING.dMinPitch = FC.PIDS[1][2];
+                            FC.ADVANCED_TUNING.dMinYaw = FC.PIDS[2][2];
                         } else {
-                            FC.PIDs[0][2] = ADVANCED_TUNING.dMinRoll;
-                            FC.PIDs[1][2] = ADVANCED_TUNING.dMinPitch;
-                            FC.PIDs[2][2] = ADVANCED_TUNING.dMinYaw;
+                            FC.PIDS[0][2] = FC.ADVANCED_TUNING.dMinRoll;
+                            FC.PIDS[1][2] = FC.ADVANCED_TUNING.dMinPitch;
+                            FC.PIDS[2][2] = FC.ADVANCED_TUNING.dMinYaw;
                         }
                         TuningSliders.calculateNewPids();
                     }
@@ -1834,10 +1837,12 @@ TABS.pid_tuning.initialize = function (callback) {
             $('input[id="useIntegratedYaw"]').change(() => TuningSliders.updatePidSlidersDisplay());
 
             // trigger Slider Display update when PID mode is changed
-            $('select[id="sliderPidsModeSelect"]').change(function () {
-                TuningSliders.sliderPidsMode = parseInt($(this).val());
-                TuningSliders.updatePidSlidersDisplay();
-            }).change();
+            if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
+                $('select[id="sliderPidsModeSelect"]').change(function () {
+                    TuningSliders.sliderPidsMode = parseInt($(this).val());
+                    TuningSliders.updatePidSlidersDisplay();
+                }).change();
+            }
 
             let allPidTuningSliders;
             if (semver.lt(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
```

**Review by haslinghuis — APPROVED** [review]
Finally. It working!!!

**Comment by haslinghuis** [comment]
if pidslidersmodeSelect is OFF sliders disable in firmware mode - maybe this is intentional.
Second. Testing now with the firmware in #10509. Seems working.
One remark: after doing `defaults` in cli the firmware sliders are disabled.

**Comment by spatzengr** [comment]
@nangel0 I'm not getting that behavior.  Maybe @haslinghuis can see what he gets.

This is the latest build from PR (minus one bug fix I need to commit) and the latest BF Master.
VZx4yjgmT7

**Comment by haslinghuis** [comment]
@nangel0 Be sure you on the latest development firmware and do a `defaults` in cli to test.
After `defaults` I still have to enable the sliders, but after that the Dmin switch in RP/RPY mode is working.
OFF mode sets the sliders disabled again.
Sliders warnings are there for all sliders when pushing the limits, but sometimes it just ignores the warning, hard to reproduce now they are all showing the warning.

**Comment by spatzengr** [comment]
@haslinghuis 
_After defaults, I still have to enable the sliders._
That is because the Default in Betaflight for `sliderPIDsModeSelect` is set to OFF.  Changing that on the FW side will make sliders enabled by Default (as intended).  That's a PR I will do on BF after this is Merged.  

If you set D-min Gain slider to 2.0, it makes D_min = D_max which disables D_min as an intended action.  I-term slider does not affect this behavior.

The behavior of warning disappearing with the above slider settings is intended.  See gain values.  They fall within the range of defaults that were deemed "safe" back in 4.1 when sliders were introduced.

File: src/js/TuningSliders.js
Comment (spatzengr) [reply src/js/TuningSliders.js]: done

**Review by McGiverGim — APPROVED** [review]

**Review by haslinghuis — APPROVED** [review]

**Comment by spatzengr** [comment]
I'm seeing two approvals.  @McGiverGim anything else I need to do here yet?  Needs another Maintiner approval?

**Comment by McGiverGim** [comment]
It's ok to me. But usually the only one that does the merge action is @mikeller

**Review by mikeller — APPROVED** [review]

**Review by mikeller — APPROVED** [review]

**Pull request merged**
